src/state/StateClient.cpp

```diff
@@ -1,138 +1,193 @@
                          const std::string& hostIn)
-  : user(userIn)
+  : faabric::transport::MessageEndpointClient(hostIn, STATE_PORT)
+  , user(userIn)
   , key(keyIn)
   , host(hostIn)
   , reg(state::getInMemoryStateRegistry())
-  , channel(grpc::CreateCustomChannel(host + ":" + std::to_string(STATE_PORT),
-                                      grpc::InsecureChannelCredentials(),
-                                      getChannelArgs()))
-  , stub(faabric::StateRPCService::NewStub(channel))
-{}
+{
+    this->open(faabric::transport::getGlobalMessageContext(),
+               faabric::transport::SocketType::PUSH,
+               false);
+}
 
-void StateClient::pushChunks(const std::vector<StateChunk>& chunks)
+StateClient::~StateClient()
 {
-    faabric::StateResponse response;
-    ClientContext clientContext;
-    auto stream = stub->Push(&clientContext, &response);
+    close();
+}
+
+void StateClient::sendHeader(faabric::state::StateCalls call)
+{
+    // Deliberately using heap allocation, so that ZeroMQ can use zero-copy
+    int functionNum = static_cast<int>(call);
+    size_t headerSize = sizeof(faabric::state::StateCalls);
+    char* header = new char[headerSize];
+    memcpy(header, &functionNum, headerSize);
+    // Mark that we are sending more messages
+    send(header, headerSize, true);
+}
 
-    for (auto chunk : chunks) {
-        faabric::StatePart c;
-        c.set_user(user);
-        c.set_key(key);
-        c.set_offset(chunk.offset);
-        c.set_data(chunk.data, chunk.length);
+// Block until call finishes, but ignore response
+void StateClient::awaitResponse()
+{
+    char* data;
+    int size;
+    awaitResponse(data, size);
+    delete data;
+}
 
-        if (!stream->Write(c)) {
-            break;
-        };
+void StateClient::awaitResponse(char*& data, int& size)
+{
+    // Call the superclass implementation
+    MessageEndpointClient::awaitResponse(
+      faabric::util::getSystemConfig().endpointHost,
+      STATE_PORT + REPLY_PORT_OFFSET,
+      data,
+      size);
+}
+
+void StateClient::sendStateRequest(bool expectReply)
+{
+    sendStateRequest(nullptr, 0, expectReply);
+}
+
+void StateClient::sendStateRequest(const uint8_t* data,
+                                   int length,
+                                   bool expectReply)
+{
+    faabric::StateRequest request;
+    request.set_user(user);
+    request.set_key(key);
+    if (length > 0) {
+        request.set_data(data, length);
     }
+    request.set_returnhost(faabric::util::getSystemConfig().endpointHost);
+    size_t requestSize = request.ByteSizeLong();
+    char* serialisedMsg = new char[requestSize];
+    // Serialise using protobuf
+    if (!request.SerializeToArray(serialisedMsg, requestSize)) {
+        throw std::runtime_error("Error serialising message");
+    }
+    send(serialisedMsg, requestSize);
+}
 
-    stream->WritesDone();
+void StateClient::pushChunks(const std::vector<StateChunk>& chunks)
```

Comment:
Note that `pushChunks` and `pullChunks` used to be streaming RPCs, whereas now we send as many messages as chunks, and acknowledge them _one by one_.